test/system/src/main/java/io/pravega/test/system/framework/services/MarathonBasedService.java

```diff
@@ -197,4 +198,27 @@ private boolean isDeploymentPresent(final String deploymentID) {
                 () -> FutureHelpers.delayedFuture(Duration.ofSeconds(5), executorService),
                 executorService);
     }
+
+    private boolean isAppBeingStaged(final String appID) {
+        try {
+            GetAppResponse app = marathonClient.getApp(this.id);
+            int stagedTaskCount = app.getApp().getTasksStaged();
+            log.debug("Number of tasks being staged for app {} is {}", appID, stagedTaskCount);
+            return stagedTaskCount != 0;
+        } catch (MarathonException ex) {
+            if (ex.getStatus() == NOT_FOUND.code()) {
+                log.info("App is not present : {}", this.id);
+                return false;
+            }
+            throw new TestFrameworkException(RequestFailed, "Marathon Exception while fetching app details", ex);
+        }
+    }
+
+    private void waitUntilStagingComplete() {
+        final CompletableFuture<Void> status = FutureHelpers.loop(() -> isAppBeingStaged(this.id), //condition
```

Comment:
is it possible to have a pre-staged state where this loop condition could fail?

Reply:
No. As discussed, the moment we create a task, the first state is STAGING.